File: Medium/longest_common_subsequence/longest_common_subsequence.py
from typing import List 

def longest_common_subsequence(text1: str, text2: str) -> int:
    m = len(text1)
    n = len(text2)

    def lcs_recursive(i, j):
        if i==0 or j==0: #If either string is empty
            return 0
        if text1[i-1] == text2[j-1]: #If last characters match, add 1 to result and recurse for remaining strings
            return 1 + lcs_recursive(i-1, j-1)
        else: #If last characters don't match, recurse by excluding last character from either string
            return max(lcs_recursive(i-1, j), lcs_recursive(i, j-1))
        
    return lcs_recursive(m, n)

if __name__ == "__main__":
    print("Testing longest_common_subsequence function:")
    print("For the input text1 = 'abcde', text2 = 'ace', the output is:", longest_common_subsequence("abcde", "ace")) #Output: 3
    print("For the input text1 = 'abc', text2 = 'abc', the output is:", longest_common_subsequence("abc", "abc")) #Output: 3
    print("For the input text1 = 'abc', text2 = 'def', the output is:", longest_common_subsequence("abc", "def")) #Output: 0

File: Medium/longest_common_subsequence/longest_common_subsequence.cpp
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    int longestCommonSubsequence(string text1, string text2) {
        int lcs = 0;
        int i = 0, j = 0;
        int m = text1.size(), n = text2.size();
        while(i < m && j < n){
            if(text1[i] == text2[j]){
                lcs++;
                i++;
                j++;
            } else{
                int temp_j = j;
                bool found = false;
                while(temp_j < n){
                    if(text1[i] == text2[temp_j]){
                        found = true;
                        break;
                    }
                    temp_j++;
                }
                if(!found){
                    i++;
                }
            }
        }
        return lcs;
    }
};

int main(){
    Solution solution1;
    string text1 = "abcde";
    string text2 = "ace";
    int result1 = solution1.longestCommonSubsequence(text1, text2);
    cout << "Longest Common Subsequence of \"" << text1 << "\" and \"" << text2 << "\" is: " << result1 << endl;

    text1 = "abc";
    text2 = "abc";
    int result2 = solution1.longestCommonSubsequence(text1, text2);
    cout << "Longest Common Subsequence of \"" << text1 << "\" and \"" << text2 << "\" is: " << result2 << endl;

    text1 = "abc";
    text2 = "def";
    int result3 = solution1.longestCommonSubsequence(text1, text2);
    cout << "Longest Common Subsequence of \"" << text1 << "\" and \"" << text2 << "\" is: " << result3 << endl;
    return 0;

}